jenkins-job-components/src/main/java/com/ofg/job/component/JobComponent.java

```diff
@@ -0,0 +1,20 @@
+package com.ofg.job.component;
+
+import javaposse.jobdsl.dsl.Job;
+
+/**
+ * Interface for configurations applicable on Job.
+ */
+public interface JobComponent<J extends Job> {
+
+    JobComponent<? extends Job> EMPTY = job -> {};
+    
+    default <NJ extends J> JobComponent<NJ> and(JobComponent<NJ> other) {
```

Comment:
My proposal maintains correctness of the composition by retaining the type when composing "upwards" :) hence the two methods needed. If I have a Configurer for MutliJob, there's no reason for disallowing composing it with an AbstractJobConfigurer - as long as the resulting configurer is still a Configurer<MultiJob>. The method I proposed gives exactly that :) Note that the two last examples in your comment don't compile with the method I propose (and most likely needs a differently-named method for the type-retaining super-accepting composition variant)